### ISSUE TITLE
Performance of Adam different in Mxnet 0.9.3 and 0.10.0

### ISSUE THREAD

**Issue by PramuPerera** [apache/incubator-mxnet#7295]
For bugs or installation issues, please provide the following information.
The more information you provide, the more likely people will be able to help you.

Operating System: ubuntu 14.04
Package used (Python): 2.7.6
MXNet version: 0.9.3 and 0.10.0

I get different results when I run the same piece of code in mxnet 0.9.3 and 0.10 versions.
The network converges in the former case where as in the latter, it doesn't. I assume this is due
to a issue with the Adam optimizer.

Sample output on mxnet 0.10.0 machine :

Epoch 0, iteration 0 , Training ('mse', 0.0035316271241754293)
Epoch 0, iteration 200 , Training ('mse', 0.0020998953378282082)
Epoch 0, iteration 400 , Training ('mse', 0.0020730258888818159)
Epoch 0, iteration 600 , Training ('mse', 0.0020803247674131228)
Epoch 0, iteration 800 , Training ('mse', 0.0020828090243904238)
Epoch 0, iteration 1000 , Training ('mse', 0.0020948627337508208)
Epoch 0, iteration 1200 , Training ('mse', 0.0020971355235109494)
Epoch 0, iteration 1400 , Training ('mse', 0.0021050738676455798)
Epoch 0, iteration 1600 , Training ('mse', 0.002108772054051679)
Epoch 0, iteration 1800 , Training ('mse', 0.0021108072839298806)
Epoch 0, iteration 2000 , Training ('mse', 0.0021140645785808288)
Epoch 0, iteration 2200 , Training ('mse', 0.0021199630068802107)
Epoch 0, iteration 2400 , Training ('mse', 0.0021227942717597912)

Sample output on mxnet 0.9.3 machine:
Epoch 0, iteration 0 , Training ('mse', 0.0035316271241754293)
Epoch 0, iteration 200 , Training ('mse', 0.0020357351889138792)
Epoch 0, iteration 400 , Training ('mse', 0.0019829668106971089)
Epoch 0, iteration 600 , Training ('mse', 0.0019657226051196159)
Epoch 0, iteration 800 , Training ('mse', 0.0019437202295176823)
Epoch 0, iteration 1000 , Training ('mse', 0.0019294929921628906)
Epoch 0, iteration 1200 , Training ('mse', 0.0019097013292767473)
Epoch 0, iteration 1400 , Training ('mse', 0.0018983960284955858)
Epoch 0, iteration 1600 , Training ('mse', 0.0018860451705657239)
Epoch 0, iteration 1800 , Training ('mse', 0.0018750134976254957)
Epoch 0, iteration 2000 , Training ('mse', 0.0018656182529947914)
Epoch 0, iteration 2200 , Training ('mse', 0.0018591857366913381)
Epoch 0, iteration 2400 , Training ('mse', 0.0018510462387049122)

I'm assuming the inconsistency is with the optimizer (I use Adam), as both networks
report identical losses at the initial iteration. When I use SGD I get almost identical
outputs.

Output obtained with SGD is as follows:

mxnet 0.9.3 machine:
Epoch 0, iteration 0 , Training ('mse', 0.0035316271241754293)
Epoch 0, iteration 200 , Training ('mse', 0.0025281838178449305)
Epoch 0, iteration 400 , Training ('mse', 0.0024361704746748963)
Epoch 0, iteration 600 , Training ('mse', 0.0024052463359929338)
Epoch 0, iteration 800 , Training ('mse', 0.0023750611564921174)
Epoch 0, iteration 1000 , Training ('mse', 0.0023559516919851231)
Epoch 0, iteration 1200 , Training ('mse', 0.0023335584792751295)
Epoch 0, iteration 1400 , Training ('mse', 0.0023223736288017997)
Epoch 0, iteration 1600 , Training ('mse', 0.0023103310963776495)

mxnet 0.10.0 machine:
Epoch 0, iteration 0 , Training ('mse', 0.0035316271241754293)
Epoch 0, iteration 200 , Training ('mse', 0.0025281838647585675)
Epoch 0, iteration 400 , Training ('mse', 0.0024361706192505329)
Epoch 0, iteration 600 , Training ('mse', 0.0024052464653863364)
Epoch 0, iteration 800 , Training ('mse', 0.0023750612553216041)
Epoch 0, iteration 1000 , Training ('mse', 0.0023559517588570612)
Epoch 0, iteration 1200 , Training ('mse', 0.0023335585225067977)
Epoch 0, iteration 1400 , Training ('mse', 0.002322373659214368)
Epoch 0, iteration 1600 , Training ('mse', 0.002310331112374758)

The two lines of codes used for SGD and Adam are as follows:
opt = mx.optimizer.SGD(momentum = 0.9, learning_rate=lr, wd=0.0001,
                        lr_scheduler=lr_sch,rescale_grad=1.0/batchsize)
opt = mx.optimizer.Adam(learning_rate=lr, wd=0.0001, lr_scheduler=lr_sch,
                        rescale_grad=1.0/batchsize)

Some of the symbols I hope to use in my code is not available in mxnet 0.93.
Therefore I'm looking at getting my code running on mxnet 0.10. At the same time,
I need to use Adam for optimization. What changes should I
make in my code to get the same performance in 0.10 as in 0.93?

**Comment by piiswrong** [comment]
This is due to a bug in adam optimizer was fixed between 0.9.3 and 0.10.0. You can try reducing weight decay to improve convergence

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!
Also, do please check out our [forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) for general "how-to" questions.
